server/embed/etcd.go

```diff
@@ -388,6 +399,24 @@ func (e *Etcd) Config() Config {
 // Close gracefully shuts down all servers/listeners.
 // Client requests will be terminated with request timeout.
 // After timeout, enforce remaning requests be closed immediately.
+//
+// The rough workflow to shut down etcd:
+//  1. close the `stopc` channel, so that all error handlers (child
+//     goroutines) won't send back any errors anymore;
+//  2. close all client and metrics listeners, so that etcd server
+//     stops receiving any new connection immediately;
+//  3. stop the http and grpc servers gracefully, within request timeout;
+//  4. call the cancel function to close the gateway context, so that
```

Comment:
Should we close gateway before we close grpc servers? Correct me if I'm wrong, gateway exposes HTTP API and translates HTTP requests to grpc API by connecting to grpc server.

Reply:
> I would prefer to revert the #19139 and not backport more changes

#19139 fixed the deadlock issue, and that part makes sense.  But the problem is that it introduced an regression due the `wg`.  We just need to fix it. 



> It's too risky break etcd just because of it.

We previously underestimated this issue, and there was also negligence when reviewing #19139. This is why I have spent so much time personally addressing the problem.